### PR TITLE
Associative array and red black trees

### PR DESCRIPTION
**Pull request by NoraRo**

### CODE REVIEW

File: cpprelude/inc/cpprelude/associative_array.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		{
+		}
+
+		associative_array(const associative_array<key_type, value_type>& other, const AllocatorT& allocator)
+			:_tree(other._tree, allocator)
+		{
+		}
+
+		rb_iterator
+		insert(const key_type& k)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: where's r-values boilerplate code?

File: cpprelude/inc/cpprelude/associative_array.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+		const value_type&
+		at(const key_type& k) const
+		{
+			pair_node n(k);
+			return _tree[n]->data.value;
+		}
+
+		value_type&
+		operator[](const key_type& key)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: const operator[] would be nice

File: cpprelude/inc/cpprelude/iterator.h [hunk 3]
```diff
@@ -403,6 +472,75 @@ namespace cpprelude
 	};
 
 	template<typename T>
+	struct rb_tree_iterator
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/iterator.h]: this means i can't iterate through this data structure ?

File: cpprelude/inc/cpprelude/associative_array.h [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+#include "cpprelude/allocator.h"
+#include "cpprelude/iterator.h"
+#include "cpprelude/rb_tree.h"
+
+namespace cpprelude
+{
+	template<typename key_type, typename value_type, typename AllocatorT = global_allocator>
+	struct associative_array
+	{
+		using pair_node = details::pair_node<key_type, value_type>;
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: add data_type

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 5]
```diff
@@ -0,0 +1,893 @@
...
+		{
+			auto it = list.begin();
+			for (usize i = 0; i < _count; ++i)
+			{
+				insert(*it);
+				it = std::next(it);
+			}
+		}
+
+		rb_tree(const rb_tree<T>& other)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: plz implement move

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 6]
```diff
@@ -0,0 +1,893 @@
...
+		operator=(const rb_tree<T>& other)
+		{
+			clear();
+			_allocator = other._allocator;
+			_copy_content(other);
+			return *this;
+		}
+
+		void 
+		_free_mem(rb_iterator it)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: put underscore functions at the end of the struct

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 7]
```diff
@@ -0,0 +1,893 @@
...
+
+		rb_iterator
+		operator[](const T& key)
+		{	//this function either returns the existed node or the new inserted node
+			return insert(key);
+		}
+
+		template<typename function_type>
+		void 
+		inorder_traverse(function_type FT)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: good stuff right there 👍

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 8]
```diff
@@ -0,0 +1,893 @@
...
+
+		template<typename function_type>
+		void
+		preorder_traverse(function_type FT)
+		{
+			_preorder_traverse(FT, _root);
+		}
+
+		void
+		_recolor(rb_iterator it)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: DAMN that's a lot of code

File: cpprelude/inc/cpprelude/iterator.h (hunk 3 above)
Comment (NoraRo) [inline cpprelude/inc/cpprelude/iterator.h]: I can by using move_up , move_left , move_right.

File: cpprelude/inc/cpprelude/rb_tree.h (hunk 7 above)
Comment (NoraRo) [inline cpprelude/inc/cpprelude/rb_tree.h]: Arigato D:

File: cpprelude/inc/cpprelude/rb_tree.h (hunk 8 above)
Comment (NoraRo) [inline cpprelude/inc/cpprelude/rb_tree.h]: eh2 eh2..

File: cpprelude/inc/cpprelude/rb_tree.h (hunk 6 above)
Comment (NoraRo) [inline cpprelude/inc/cpprelude/rb_tree.h]: Done

File: cpprelude/inc/cpprelude/associative_array.h (hunk 4 above)
Comment (NoraRo) [inline cpprelude/inc/cpprelude/associative_array.h]: here we have two data types one for the key and one for the value which one should the data_type be? or make it pair_node?

File: cpprelude/inc/cpprelude/associative_array.h (hunk 1 above)
Comment (NoraRo) [inline cpprelude/inc/cpprelude/associative_array.h]: Done, but needs lil review again.

File: cpprelude/inc/cpprelude/associative_array.h (hunk 4 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: mmmm key_type, value_type and also data_type as pair_node

File: cpprelude/inc/cpprelude/iterator.h (hunk 3 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/iterator.h]: no what i mean is that i cannot do this
```
for(auto entry: associative_array)
```

File: cpprelude/inc/cpprelude/associative_array.h [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+		
+		associative_array(const AllocatorT& allocator = AllocatorT())
+			:_tree(allocator)
+		{}
+
+		associative_array(std::initializer_list<pair_node> list, const AllocatorT& allocator = AllocatorT())
+			: _tree(allocator)
+		{
+			auto it = list.begin();
+			for (usize i = 0; i < _count; ++i)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: change _count to list.size()

File: cpprelude/inc/cpprelude/associative_array.h [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+			{
+				tree.insert(*it);
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: change tree to _tree

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 11]
```diff
@@ -0,0 +1,983 @@
...
+			if (_root == nullptr)
+			{
+				_root = &(*it_new_node);
+				_recolor(it_new_node);
+				++_count;
+				return it_new_node;
+			}
+			else
+			{
+				auto temp = _insert(_root, tmp::move(*it_new_node));
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: you shouldn't move the iterator

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 12]
```diff
@@ -0,0 +1,983 @@
...
+				_reset_root();
+				++_count;
+				return temp;
+			}
+		}
+
+		void
+		delete_rb_tree(const T&k)
+		{
+			RB_Node* node_to_delete = &(*lookup(k));
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: why not use `lookup(k)->_node` directly

File: cpprelude/inc/cpprelude/associative_array.h [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+			return _tree[n]->data.value;
+		}
+
+		value_type&
+		operator[](const key_type& key)
+		{
+			return at(key);
+		}
+
+		value_type
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/associative_array.h]: const value_type&

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 14]
```diff
@@ -0,0 +1,983 @@
...
+
+		rb_iterator
+		operator[](const T& key)
+		{	//this function either returns the existed node or the new inserted node
+			return insert(key);
+		}
+
+		template<typename function_type>
+		void
+		inorder_traverse(function_type FT)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: expose this function to associative_array

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 15]
```diff
@@ -0,0 +1,983 @@
...
+		{
+			_inorder_traverse(FT, _root);
+		}
+
+		template<typename function_type>
+		void
+		postorder_traverse(function_type FT)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: expose this function to associative_array

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 16]
```diff
@@ -0,0 +1,983 @@
...
+		{
+			_postorder_traverse(FT, _root);
+		}
+
+		template<typename function_type>
+		void
+		preorder_traverse(function_type FT)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: expose this function to associative_array

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 17]
```diff
@@ -0,0 +1,983 @@
...
+			tmp::swap(_root, other._root);
+			tmp::swap(_count, other._count);
+			tmp::swap(_allocator, other._allocator);
+		}
+
+		void
+		swap(rb_tree&& other)
+		{
+			tmp::swap(_root, other._root);
+			tmp::swap(_count, tmp::move(other._count));
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: move is not worth using on a number

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 18]
```diff
@@ -0,0 +1,983 @@
...
+			tmp::swap(_allocator, tmp::move(other._allocator));
+		}
+
+		rb_iterator
+		get_min() const
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: name it just `min`

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 19]
```diff
@@ -0,0 +1,983 @@
...
+
+			while (min->left != nullptr)
+			{
+				min = min->left;
+			}
+			return min;
+		}
+
+		rb_iterator
+		get_max() const
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: `max`

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 20]
```diff
@@ -0,0 +1,983 @@
...
+
+			while (max->right != nullptr)
+			{
+				max = max->right;
+			}
+			return max;
+		}
+
+		rb_iterator
+		get_root() const
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: `root`

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 21]
```diff
@@ -0,0 +1,983 @@
...
+			return (-1 != _calculate_Black_count(_root));
+		}
+
+		void
+		_free_mem(rb_iterator it)
+		{
+			if (it == nullptr) return;
+
+			it->data.~T();
+			_allocator.free(make_slice(&(*it)));
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: same `it->_node`

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 22]
```diff
@@ -0,0 +1,983 @@
...
+				if (it == nullptr) continue;
+				
+				insert(it->data);
+				queue.enqueue(it->left);
+				queue.enqueue(it->right);
+			}
+		}
+
+		void
+		_copy_content(rb_tree<T>&& other)
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: why copy content works the same on an r-value tree ... the user says to you "you can destruct this object i don't care about it anymore" so i think the appropriate thing to do is just steal its guts

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 23]
```diff
@@ -0,0 +1,983 @@
...
+
+			//if parent is REd then it has grandpa
+			bool is_parent_on_left_branch = it->parent->parent->data > it->parent->data;
+
+			//is the parent node on the left branch from grandparent node or 
+			//on the right branch.
+			if (is_parent_on_left_branch)
+			{
+				auto grandpa_node = it->parent->parent;
+				auto uncle_node = grandpa_node->right;
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: lol the names

File: cpprelude/inc/cpprelude/rb_tree.h [hunk 24]
```diff
@@ -0,0 +1,983 @@
...
+			auto temp = _allocator.template alloc<RB_Node>();
+			new (temp) RB_Node(k);
+			return rb_iterator(temp);
+		}
+
+		rb_iterator
+		_create_node(T&& k)
+		{
+			auto temp = _allocator.template alloc<RB_Node>();
+			new (temp) RB_Node(k);
```
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/rb_tree.h]: tmp::move(k)

**Pull request closed without merging**